### PR TITLE
Add suspend feature to the bot.

### PR DESCRIPTION
**Pull request by BrightSkyz**
This adds the suspend to the bot.

### CODE REVIEW

File: src/command/DiscordCommand.ts [hunk 1]
```diff
@@ -1,17 +1,19 @@
...
 
 export abstract class DiscordCommand {
   protected readonly dependencies: CommandDependencies;
   protected readonly command: string;
   protected readonly args: Array<string>;
   protected readonly message: Message;
+  protected readonly messageChannel: TextChannel;
 
-  public constructor(dependencies: CommandDependencies, command: string, args: Array<string>, message: Message) {
+  public constructor(dependencies: CommandDependencies, command: string, args: Array<string>, message: Message, messageChannel: TextChannel) {
```
Comment (COVID-GUH) [inline src/command/DiscordCommand.ts]: Why is it necessary to store this? Not all commands need the channel, hence modifying the base class is not great. I would consider further inheritance if this is absolutely required. Can you explain why?

File: src/command/DiscordCommandListener.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export class DiscordCommandListener {
   private async handleDiscordCommand(message: Message): Promise<void> {
     if (!message.content.startsWith(DISCORD_PREFIX!) || message.author.bot || message.channel.type !== "text") return;
 
-    const args: Array<string> = message.content?.slice(DISCORD_PREFIX!.length).split(" ");
+    const args: Array<string> = message.content?.slice(DISCORD_PREFIX!.length).split(" ").filter((arg) => arg.length !== 0);
```
Comment (COVID-GUH) [inline src/command/DiscordCommandListener.ts]: Why is this necessary?

File: src/command/DiscordCommandRegistry.ts [hunk 3]
```diff
@@ -1,16 +1,23 @@
...
 
-import { Message } from "discord.js";
-import { HelpDiscordCommand } from "./help/HelpDiscordCommand";
+import {Message, TextChannel} from "discord.js";
 
 export class DiscordCommandRegistry {
   private static getRegistry(): Map<string, DiscordCommand> {
     const registry = new Map<DiscordCommandType, DiscordCommand>();
+    registry.set(DiscordCommandType.HISTORY, HistoryDiscordCommand.prototype);
+    registry.set(DiscordCommandType.STUCKCHANNEL, StuckChannelDiscordCommand.prototype);
```
Comment (COVID-GUH) [inline src/command/DiscordCommandRegistry.ts]: snake case please

File: src/database/QuarantineRepository.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    return result.rows[0]["quarantine_id"];
+  }
+
+  public async getMostRecentByOffenderDiscordId(offenderDiscordId: string): Promise<Quarantine | null> {
+    const statement = "SELECT q.* FROM quarantines q INNER JOIN users u on q.offender_user_id = u.user_id WHERE u.discord_id = $1 ORDER BY q.created_at DESC LIMIT 1";
+    const result = await this.postgresDriver.query(statement, [offenderDiscordId]);
+    return result.rowCount > 0 ? result.rows[0] : null;
+  }
+
+  public async getFiveMostRecentByOffenderDiscordId(offenderDiscordId: string): Promise<Array<Quarantine>> {
```
Comment (COVID-GUH) [inline src/database/QuarantineRepository.ts]: This is too contrived, we want to be more generic and deal with this programmatically.

File: src/command/punishment/SuspendDiscordCommand.ts [hunk 5]
```diff
@@ -1,13 +1,115 @@
 import { DiscordCommand } from "../DiscordCommand";
-import { MessageEmbed } from "discord.js";
+import {Channel, GuildMember, MessageEmbed} from "discord.js";
+
+const { DISCORD_PREFIX, QUARANTINE_ROLES, STAFF_ROLES } = process.env;
+
+const REPLACE_MENTION_REGEX = /[<>!@]/g;
```
Comment (freememory) [inline src/command/punishment/SuspendDiscordCommand.ts]: The REPLACE_MENTION_REGEX is used in several places to do the exact same thing. Refactor.

File: src/command/punishment/SuspendDiscordCommand.ts [hunk 6]
```diff
@@ -1,13 +1,115 @@
...
+        type: "member",
+        allow: ["VIEW_CHANNEL"],
+        id: qtCategory.id
+      }],
+      parent: qtCategory.id
+    });
+
+    if (qtChannel) {
+      await qtChannel.send({
+        embed: new MessageEmbed().setTitle("Suspend").setFooter("Uh-oh.").setDescription("You have been suspeneded, <@" + member.id + ">.\nReason: " + (reasonProvided ? reasonProvided : "None"))
```
Comment (freememory) [inline src/command/punishment/SuspendDiscordCommand.ts]: Pet peeve of mine personally. I seriously dislike the "Uh oh" footer in the suspend message. It makes it feel very juvenile. I'll back off on it if everyone else likes it, but I personally find it to be juvenile.

File: src/command/punishment/UnsuspendDiscordCommand.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    if (quarantine) {
+      const foundChannel = guild?.channels.cache.filter(channel => channel.id === quarantine.channel_id).first();
+      if (foundChannel) {
+        foundChannel.delete("Channel cleanup from quarantine.");
+      }
+    }
+  }
+
+  public async validate(): Promise<boolean> {
+    if (!this.message.member?.roles.cache.some(role => AUTHORIZED_ROLES.includes(role.name))) {
```
Comment (freememory) [inline src/command/punishment/UnsuspendDiscordCommand.ts]: This logic, also, appears everywhere. A standard validation for all commands is:
* can I run this?
* can you run this?

Can this also be refactored?

Comment (BrightSkyz) [reply src/command/punishment/UnsuspendDiscordCommand.ts]: No, this cannot be changed, it needs to be there due to the fact there will be some commands that anyone can run, no matter the role. This just allows the people with the role listed in the config to use the quarantine commands specifically.

Comment (COVID-GUH) [reply src/command/punishment/UnsuspendDiscordCommand.ts]: I agree with @freememory, there seems to be some disconnect on the usage of this pattern.

File: src/database/QuarantineRepository.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  public async getMostRecentByOffenderDiscordId(offenderDiscordId: string): Promise<Quarantine | null> {
+    const statement = "SELECT q.* FROM quarantines q INNER JOIN users u on q.offender_user_id = u.user_id WHERE u.discord_id = $1 ORDER BY q.created_at DESC LIMIT 1";
+    const result = await this.postgresDriver.query(statement, [offenderDiscordId]);
+    return result.rowCount > 0 ? result.rows[0] : null;
+  }
+
+  public async getMultipleMostRecentByOffenderDiscordId(offenderDiscordId: string, amountToFetch: number): Promise<Array<Quarantine>> {
+    const statement = "SELECT q.* FROM quarantines q INNER JOIN users u on q.offender_user_id = u.user_id WHERE u.discord_id = $1 ORDER BY q.created_at DESC LIMIT " + amountToFetch;
```
Comment (COVID-GUH) [inline src/database/QuarantineRepository.ts]: This is opening a hole for SQL injection. I will fix this.
